pyzeebe/worker/worker.py

```diff
@@ -68,19 +90,83 @@ def stop(self, wait: bool = False) -> None:
         """
         self.stop_event.set()
         if wait:
-            logger.debug("Waiting for threads to join")
-            while self._task_threads:
-                thread = self._task_threads.pop(0)
-                thread.join()
+            self._join_task_threads()
+
+    def _join_task_threads(self) -> None:
+        logger.debug("Waiting for threads to join")
+        while self._task_threads:
+            _, thread = self._task_threads.popitem()
+            thread.join()
+        logger.debug("All threads joined")
+
+    def _watch_task_threads(self, frequency: int = 10) -> None:
+        logger.debug("Starting task thread watch")
+        try:
+            self._watch_task_threads_runner(frequency)
+        except Exception as err:
+            if isinstance(err, MaxConsecutiveTaskThread):
+                logger.debug("Stopping worker due to too many errors.")
+            else:
+                logger.debug("An unhandled exception occured when watching threads, stopping worker")
+            self.stop()
+            raise
+        logger.info(f"Watcher stopping (stop_event={self.stop_event.is_set()}, "
+                    f"task_threads list lenght={len(self._task_threads)})")
+
+    def _should_handle_task(self) -> bool:
+        return not self.stop_event.is_set() and (self.zeebe_adapter.connected or self.zeebe_adapter.retrying_connection)
+
+    def _should_watch_threads(self) -> bool:
+        return not self.stop_event.is_set() and bool(self._task_threads)
+
+    def _watch_task_threads_runner(self, frequency: int = 10) -> None:
+        consecutive_errors = 0
+        max_errors = self.watcher_max_errors_factor * len(self.tasks)
+        while self._should_watch_threads():
+            logger.debug("Checking task thread status")
+            # converting to list to avoid "RuntimeError: dictionary changed size during iteration"
+            for task_type in list(self._task_threads.keys()):
+                # thread might be none, if dict changed size, in that case we'll consider it
+                # an error, and check if we should handle it
+                thread = self._task_threads.get(task_type)
+                if not thread or not thread.is_alive():
+                    consecutive_errors += 1
+                    self._check_max_errors(consecutive_errors, max_errors)
+                    self._handle_not_alive_thread(task_type)
+                else:
+                    consecutive_errors = 0
+            time.sleep(frequency)
+
+    def _handle_not_alive_thread(self, task_type):
```

Comment:
```suggestion
    def _handle_not_alive_thread(self, task_type: str):
```

Reply:
Addressed in 945846941c92ee64afc8044bce8f9a5407f013b8